### DISCUSSION TITLE
Refactoring full page component to parent + children?

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#6074] [1. Help]
We have a full page component that at its core has 3 separate sections:

- Quote
- Product (morph relation with Quote)
- Pricing (has many morph relation with Quote)

At the moment they are all snug in a single Livewire component, which I don't particularly like as it has a lot of responsibility. Also, the product is dynamic. The fields and validation change depending on which product the quote is for. The quote needs to be stored before the other two segments and once everything is saved dispatch a Laravel event.

I tried to refactor this, but no matter what I try, it always ends up being similar to what I have now, just a little different.

What would be a good approach here?

ps. for those wondering, we use morph as we have different components setup similarly.

**Comment by joshhanley** [comment]
@ju5t that's really hard to say without knowing exactly how the component and all the sections work. If they are all forms working on the same model (ie quote) then sometimes it's easier to keep it all in one component. You could make it more maintainable in V3 by making use of FormObjects.

**Reply by ju5t** [reply]
All sections relate to the quote, but each is stored in its own model/table. For clarity, they are all in a single form.

I've drafted a solution where we're using data objects, with Spatie's package. We're storing Livewire state in simple arrays. Our `save`-method will convert the Livewire state into data objects, which we will then pass onto an action-class to process it further.

A data object contains the state and the validation rules. I've extended the data objects with a method to build validators, like full `Validator::make()`-validators. Each validator can add the prefix of the section to the error messages. This allows us to have a validation rule of `'customer_id' => ['required']` in our data object, but the error message will be prefixed with the property on the Livewire component, e.g. `quote.customer_id.required`.

The action runs the validation in series, checks for any errors and throws a new `ValidationException::withMessages()`, containing all the rewritten error messages.
